Expose model entity transform to JS

eddydas opened the pull request:
#### f0845c44e6b01ab1d954725d84b6423ff499d075
<pre>
Expose model entity transform to JS
<a href="https://rdar.apple.com/124986396">rdar://124986396</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=271208">https://bugs.webkit.org/show_bug.cgi?id=271208</a>

Reviewed by Alex Christensen.

Exposes the root entity transform matrix to JS side, so that developers
can move the 3D object around within the &lt;model&gt; tag portal.

HTMLModelTag has an m_entityTransform &quot;truth&quot; that&apos;s used to serve JS
getter synchronously.

ModelPlayer now has the ability to determine whether a certain transform
matrix is supported. If not, HTMLModelElement::setEntityTransform will
raise an Unsupported Exception.

Upon model load, Model Process will send an IPC message to Web Process
to update the m_entityTransform state of the HTMLModelElement to make
it aware of the default applied transform.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::modelPlayer const):
(WebCore::HTMLModelElement::entityTransform const):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::entityTransform):
(WebCore::ModelPlayer::setEntityTransform):
(WebCore::ModelPlayer::supportsTransform):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/platform/graphics/cocoa/TransformationMatrixCocoa.cpp:
(WebCore::TransformationMatrix::TransformationMatrix):
(WebCore::TransformationMatrix::operator simd_float4x4 const):
Added convenience converter methods to/from simd_float4x4, which is the commonly used transform matrix data type on visionOS.
Applicable only on Cocoa platforms.

* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:

(WebKit::ModelProcessModelPlayerProxy::transformSupported):
Require the transform to be a non-shear, uniform-scaling matrix, for now.

(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateTransform):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::entityTransform):
(WebKit::ModelProcessModelPlayerProxy::setEntityTransform):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didUpdateEntityTransform):
(WebKit::ModelProcessModelPlayer::entityTransform):
(WebKit::ModelProcessModelPlayer::setEntityTransform):
(WebKit::ModelProcessModelPlayer::supportsTransform):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:

Canonical link: <a href="https://commits.webkit.org/276661@main">https://commits.webkit.org/276661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4ef5cac53efb26c187b10db9d1c70102880283a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45286 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47950 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41294 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47593 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28607 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21802 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21461 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/39062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18241 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40146 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3333 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41566 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40463 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49658 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/20269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16805 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44173 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21576 "Built successfully") | | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42977 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10067 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21262 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -377,6 +382,14 @@ void HTMLModelElement::applyBackgroundColor(Color color)
+
+RefPtr<HTMLModelElementEntity> HTMLModelElement::rootEntity() const
+{
+    auto entity = HTMLModelElementEntity::create(*this);
```
Should you really return a new one every time?

eddydas replied:
Good catch. I probably shouldn't! Fixed it.

smfr commented on Source/WebCore/Modules/model-element/HTMLModelElementEntity.idl:
```diff
@@ -0,0 +1,34 @@
+
+[
+    Conditional=MODEL_ELEMENT&MODEL_PROCESS,
+    EnabledBySetting=ModelElementEnabled&ModelProcessEnabled,
```
Spaces around the `&` please on both lines.

smfr commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -127,6 +131,8 @@ class ModelProcessModelPlayerProxy final
     float m_yaw { 0 };
     REAnimationPlaybackToken m_animationPlaybackToken { kInvalidAnimationToken };
+
+    RESRT m_srt;
```
Please expand the "srt" acronym (I have no idea what it means).

eddydas replied:
Good one. Added more comments as well.
```
    RESRT m_transformSRT; // SRT=Scaling/Rotation/Translation. This is stricter than a WebCore::TransformationMatrix.
```

smfr commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -75,6 +75,12 @@ void ModelProcessModelPlayer::didFinishLoading()
 
+void ModelProcessModelPlayer::DidUpdateEntityTransform(WebCore::TransformationMatrix transform)
+{
+    RELEASE_LOG(ModelElement, "%p - ModelProcessModelPlayer DidUpdateEntityTransform id=%" PRIu64, this, m_id.toUInt64());
```
Does this really justify release logging?

eddydas replied:
Removed

smfr commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -115,6 +121,17 @@ void ModelProcessModelPlayer::setBackgroundColor(WebCore::Color color)
+
+void ModelProcessModelPlayer::setEntityTransform(WebCore::TransformationMatrix transform)
+{
+    m_transform = transform;
```
Might want to detect no-op setting.

annevk commented on Source/WebCore/Modules/model-element/HTMLModelElementEntity.cpp:
```diff
@@ -0,0 +1,88 @@
+
+    TransformationMatrix matrix = player->entityTransform();
+    return DOMMatrix::create(matrix, DOMMatrixReadOnly::Is2D::No).get();
+}
```
This ends up creating a new object each time you invoke the getter, right? That violates: https://w3ctag.github.io/design-principles/#attributes-like-data. In particular we want

> `obj.attribute === obj.attribute`

to be true.

I suspect that instead something like `setView()` and `getView()` would be better. And perhaps they can be directly on the `model` element or do we have a strong reason for this intermediate object?

eddydas replied:
Discussed offline, going to do away the whole HTMLModelElementEntity and flatten the get/setTransform into HTMLModelElement itself.

annevk commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -377,6 +382,17 @@ void HTMLModelElement::applyBackgroundColor(Color color)
         m_modelPlayer->setBackgroundColor(color);
 }
 
+// MARK: - RootElement support.
```
What does this mean?

aprotyas replied:
I think something like `#pragma mark`?

eddydas replied:
No longer relevant.

hortont424 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -72,9 +73,10 @@ using namespace HTMLNames;
 WTF_MAKE_ISO_ALLOCATED_IMPL(HTMLModelElement);
 
 HTMLModelElement::HTMLModelElement(const QualifiedName& tagName, Document& document)
-    : HTMLElement(tagName, document, { TypeFlag::HasCustomStyleResolveCallbacks, TypeFlag::HasDidMoveToNewDocument })
```
This style change is wrong

eddydas replied:
somehow the formatter wants me to do that 🤔 said "things inside a namespace should not be indented"
I guess I'll just keep it as-is and ignore the formatter

annevk commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -377,6 +388,37 @@ void HTMLModelElement::applyBackgroundColor(Color color)
+    return m_entityTransform.get();
+}
+
+ExceptionOr<const DOMMatrixReadOnly&> HTMLModelElement::setEntityTransform(const DOMMatrixReadOnly& transform)
```
This should be `ExceptionOr<void>`. A setter doesn't return anything. You'll also need to adjust wherever you attempt to return things below that are not exceptions.

eddydas replied:
Got it. Fixed! Thank you.

annevk commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -377,6 +388,37 @@ void HTMLModelElement::applyBackgroundColor(Color color)
+
+    TransformationMatrix matrix = transform.transformationMatrix();
+
+    if (!player->supportsTransform(matrix))
+        return Exception { ExceptionCode::NotSupportedError };
```
Ideally we have tentative web-platform-tests for this.

eddydas replied:
Agree. Some additional tests on the TransformationMatrix conversion and supportsTransform() is also desirable, and I'm going to add them next via rdar://125239025

annevk commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
```diff
@@ -115,6 +122,17 @@ void ModelProcessModelPlayer::setBackgroundColor(WebCore::Color color)
+    send(Messages::ModelProcessModelPlayerProxy::SetEntityTransform(transform));
+}
+
+bool ModelProcessModelPlayer::supportsTransform(WebCore::TransformationMatrix transform)
+{
+    return ModelProcessModelPlayerProxy::transformSupported(transform);
+}
```
I'm not sure I understand this. Does this mean you're doing synchronous IPC?

eddydas replied:
this is a static method. So WP side simply calls the RE lib math functions for the validation without invoking Model Process

annevk reviewed (commented):
The model element side looks good. Thanks for simplifying it and pulling the matrix validation in. Someone else should review the ModelProcess changes.

achristensen07 commented on Source/WebCore/Modules/model-element/ModelPlayer.cpp:
```diff
@@ -46,6 +51,11 @@ bool ModelPlayer::supportsDragging()
 
+bool ModelPlayer::supportsTransform(TransformationMatrix)
+{
+    return true;
```
Would false be a better default value here?

eddydas replied:
Makes sense! I'll make it false by default.

achristensen07 commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -326,6 +326,10 @@
      ["-readability/naming/underscores"]),
 
+    ([  # simd.h structs all have underscores in them.
+     os.path.join('Source', 'WebCore', 'platform', 'graphics', 'cocoa', 'TransformationMatrixCocoa.cpp')],
```
This is a bit excessive and specific.  Can't we just ignore the checker for the probably only 1 PR that will have this as an issue?

eddydas replied:
I believe we can! Will take this diff out

eddydas committed:
Expose model entity transform to JS
rdar://124986396
https://bugs.webkit.org/show_bug.cgi?id=271208

Reviewed by Alex Christensen.

Exposes the root entity transform matrix to JS side, so that developers
can move the 3D object around within the <model> tag portal.

HTMLModelTag has an m_entityTransform "truth" that's used to serve JS
getter synchronously.

ModelPlayer now has the ability to determine whether a certain transform
matrix is supported. If not, HTMLModelElement::setEntityTransform will
raise an Unsupported Exception.

Upon model load, Model Process will send an IPC message to Web Process
to update the m_entityTransform state of the HTMLModelElement to make
it aware of the default applied transform.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::modelPlayer const):
(WebCore::HTMLModelElement::entityTransform const):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::entityTransform):
(WebCore::ModelPlayer::setEntityTransform):
(WebCore::ModelPlayer::supportsTransform):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/platform/graphics/cocoa/TransformationMatrixCocoa.cpp:
(WebCore::TransformationMatrix::TransformationMatrix):
(WebCore::TransformationMatrix::operator simd_float4x4 const):
Added convenience converter methods to/from simd_float4x4, which is the commonly used transform matrix data type on visionOS.
Applicable only on Cocoa platforms.

* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:

(WebKit::ModelProcessModelPlayerProxy::transformSupported):
Require the transform to be a non-shear, uniform-scaling matrix, for now.

(WebKit::ModelProcessModelPlayerProxy::computeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateTransform):
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::entityTransform):
(WebKit::ModelProcessModelPlayerProxy::setEntityTransform):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::didUpdateEntityTransform):
(WebKit::ModelProcessModelPlayer::entityTransform):
(WebKit::ModelProcessModelPlayer::setEntityTransform):
(WebKit::ModelProcessModelPlayer::supportsTransform):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in:

Canonical link: https://commits.webkit.org/276661@main
